### ISSUE TITLE
parent_id in a form returns "undefined method"

### ISSUE THREAD

**Issue by marcamillion** [tjchambers/ancestry#1]
I am not sure if this is related to the port and the modifications, but I noticed that when I am trying to do a simple "hook-up" of a model, it won't work.

This is my class:

``` ruby
class Node < ActiveRecord::Base
  has_ancestry orphan_strategy: :adopt, cache_depth: true
```

This is my schema for my `Node.rb` table:

``` ruby
# == Schema Information
#
# Table name: nodes
#
#  id             :integer          not null, primary key
#  family_tree_id :integer
#  created_at     :datetime
#  updated_at     :datetime
#  name           :string(255)
#  ancestry       :string(255)
#  ancestry_depth :integer          default(0)
#  max_tree_depth :integer          default(0)
```

This is my controller create & new actions:

``` ruby
  def new
    @node = Node.new(parent_id: params[:parent_id])
    respond_with(@node)
  end

  def create
    @family_tree = current_user.family_tree
    @node = @family_tree.nodes.new(node_params)
    @node.save
    respond_with(@node)
  end
```

My `_form.html.erb` (which is a simple_form, for what it's worth):

`<%= f.input :parent_id, as: :hidden %>`

This is the error I am getting:

``` ruby
undefined method `parent_id' for #<Node:0x007fac361ba1e0>
```

Any ideas?

**Edit 1**

For what it's worth, I swapped out the gems in my Gemfile and it works - as is - with the stock gem.

Here are the server logs for the working version:

``` ruby
Started GET "/nodes/new?parent_id=15" for 127.0.0.1 at 2015-01-17 19:07:04 -0500
Processing by NodesController#new as HTML
  Parameters: {"parent_id"=>"15"}
  Node Load (1.4ms)  SELECT  "nodes".* FROM "nodes"  WHERE "nodes"."id" = $1 LIMIT 1  [["id", 15]]
  FamilyTree Load (2.0ms)  SELECT "family_trees".* FROM "family_trees"
  Rendered nodes/_form.html.erb (19.3ms)
  Rendered nodes/new.html.erb within layouts/application (20.6ms)
  User Load (1.8ms)  SELECT  "users".* FROM "users"  WHERE "users"."id" = 1  ORDER BY "users"."id" ASC LIMIT 1
Completed 200 OK in 337ms (Views: 328.7ms | ActiveRecord: 5.1ms)


Started POST "/nodes" for 127.0.0.1 at 2015-01-17 19:07:14 -0500
Processing by NodesController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"F1HGLwIjDG3VaIYlnsu7NbROEiWBO8xqJtjh5MreI9E=", "node"=>{"parent_id"=>"15", "name"=>"Jerry", "family_tree_id"=>"2"}, "commit"=>"Create Node"}
  User Load (2.0ms)  SELECT  "users".* FROM "users"  WHERE "users"."id" = 1  ORDER BY "users"."id" ASC LIMIT 1
  FamilyTree Load (1.3ms)  SELECT  "family_trees".* FROM "family_trees"  WHERE "family_trees"."user_id" = $1 LIMIT 1  [["user_id", 1]]
  Node Load (4.7ms)  SELECT  "nodes".* FROM "nodes"  WHERE "nodes"."id" = $1 LIMIT 1  [["id", 15]]
   (9.4ms)  BEGIN
  SQL (5.1ms)  INSERT INTO "nodes" ("ancestry", "ancestry_depth", "created_at", "family_tree_id", "name", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["ancestry", "15"], ["ancestry_depth", 1], ["created_at", "2015-01-18 00:07:14.201648"], ["family_tree_id", 2], ["name", "Jerry"], ["updated_at", "2015-01-18 00:07:14.201648"]]
   (1.4ms)  COMMIT
Redirected to http://localhost:3000/nodes/16
Completed 302 Found in 57ms (ActiveRecord: 24.0ms)


Started GET "/nodes/16" for 127.0.0.1 at 2015-01-17 19:07:14 -0500
Processing by NodesController#show as HTML
  Parameters: {"id"=>"16"}
  Node Load (4.2ms)  SELECT  "nodes".* FROM "nodes"  WHERE "nodes"."id" = $1 LIMIT 1  [["id", 16]]
  FamilyTree Load (2.6ms)  SELECT  "family_trees".* FROM "family_trees"  WHERE "family_trees"."id" = $1 LIMIT 1  [["id", 2]]
  User Load (5.0ms)  SELECT  "users".* FROM "users"  WHERE "users"."id" = $1 LIMIT 1  [["id", 1]]
  Rendered nodes/show.html.erb within layouts/application (13.3ms)
  User Load (3.2ms)  SELECT  "users".* FROM "users"  WHERE "users"."id" = 1  ORDER BY "users"."id" ASC LIMIT 1
Completed 200 OK in 274ms (Views: 256.1ms | ActiveRecord: 15.0ms)
```

This is the server log for the failed one:

``` ruby
Started GET "/nodes/new?parent_id=13" for 127.0.0.1 at 2015-01-17 18:45:27 -0500
  ActiveRecord::SchemaMigration Load (2.2ms)  SELECT "schema_migrations".* FROM "schema_migrations"
Processing by NodesController#new as HTML
  Parameters: {"parent_id"=>"13"}
  Node Load (1.4ms)  SELECT  "nodes".* FROM "nodes"  WHERE "nodes"."id" = $1 LIMIT 1  [["id", 13]]
  Rendered nodes/_form.html.erb (64.1ms)
  Rendered nodes/new.html.erb within layouts/application (70.1ms)
Completed 500 Internal Server Error in 122ms

NoMethodError - undefined method `parent_id' for #<Node:0x007fac361ba1e0>:
```

**Comment by tjchambers** [comment]
Sorry, I am at a loss. Since parent_id= is a method on a ancestry node, I am thinking for some reason it does not understand it is correct. I would look at the methods on the node before you assign it to see what they are. 

Also note that your has_ancestry config is not viable with multiple parents. The orphaned settings only set destroy as option, and cache-depth is non-useful as there is no concept of depth for multiparents. 

Sorry I can't be of more help. Let me know if you figure it out - it may be helpful to know which line on the controller is the issue and whether it can be reproduced outside of a controller.

**Comment by marcamillion** [comment]
Not sure if this helps but I am digging deeper and figured I would share, to see if you can point me in any direction.

The issue seems to be with my form.

This is the `_form.html.erb` partial:

``` erb
<%= simple_form_for(@node) do |f| %>
        <%= f.input :parent_id, as: :hidden %>

        <div class="field">
            <%= f.input :name, label: "Name", input_html: { maxlength: 20 } %>
        </div>
    <div class="field">
            <%= f.association :family_tree, label_method: :name, label: "Family Tree" %>
        </div>  
        <br />
    <div class="actions">
            <%= f.submit %>
        </div>
<% end %>
```

That generates this error:

``` error
NoMethodError at /nodes/new
undefined method `parent_id' for #<Node:0x007f951c1ec018>
```

Which corresponds to this action in the controller:

``` ruby
  def new
    @node = Node.new(parent_id: params[:parent_id])
    respond_with(@node)
  end
```

I was also digging through the logs and this line jumped out at me:

``` error
 actionpack (4.1.6) lib/action_controller/metal/mime_responds.rb:400:in `respond_with'
   () /ancestry_app/app/controllers/nodes_controller.rb:17:in `new'
```

All of this could be non-related and just be another stack trace message that is not the root cause, but figured I would just document this stuff as I come across it.

**Edit 1**

More debugging info. I have better_errors installed, and so I get a REPL on my error page - which allows me to be able to check variable values right there. When I try to get the params values, I get a weird error:

``` ruby
>> unscoped_find(params[:parent_id])
!! #<NameError: undefined local variable or method `params' for #<Node:0x007fc4db2cd678>>
>> params[:parent_id]
!! #<NameError: undefined local variable or method `params' for #<Node:0x007fc4db2cd678>>
```

Even though, right below it, it tells me this in the `Request Info` section:

``` error
Request info

Request parameters  
{"parent_id"=>"16", "action"=>"new", "controller"=>"nodes"}
```

**Comment by tjchambers** [comment]
It feels like it is due to parent_id being a method, not an attribute of the model. I would try changing your code to:

```
 def new
    @node = Node.new
    @node.parent_id = params[:parent_id]
    respond_with(@node)
  end
```

**Comment by marcamillion** [comment]
That's a great idea, but unfortunately that doesn't work. I do think it has something to do with the `params[:parent_id]` portion of the request. It feels like it may be thinking that `params` should be a method on my `Node` model, rather than an AR method.

Something is overriding params - because I can't get access to any of my params the normal way at the REPL.

**Comment by tjchambers** [comment]
So in the code have you tried printing out params hash? I see it displaying them, but am not sure they are in the params hash.

Have you looked at http://stackoverflow.com/questions/4352839/rails-how-to-get-all-parameters-from-url ? 

Not sure I am helping...

**Comment by marcamillion** [comment]
Ok I finally figured it out. When I explain it, it will be a major "duh" moment (or rather it was for me anyway, hindsight is always 20/20).

It turns out, that in your transposition to managing multiple parents, you axed the method `parent_id` altogether. 

You only have a parent setter - `parent=(parent)`, a `parent_id` setter, a `parent_ids` getter, and a `parents` getter. 

There is no getter for `parent_id`.

So I simply re-added it from the original gem and it works like a charm, i.e.:

``` ruby
    def parent_id
      if ancestor_ids.empty? then nil else ancestor_ids.last end
    end
```

I will submit a PR.....if only I can figure out how to do that from my local bundled copy of the gem. Should I clone the gem locally and make the change and submit the PR? 

I am making the changes to my locally bundled version.

**Comment by tjchambers** [comment]
I would imagine params is a Hashwithindifferentaccess but if it was a normal hash it could be the symbol not matching the string?

**Comment by marcamillion** [comment]
By the way....all of that doesn't fully explain why `params[:parent_id]` was returning that weird error.

Actually, lemme force an error and test it now - because they could be related...which would be interesting.

**Comment by tjchambers** [comment]
I am not sure why the getter is the issue here - I thought you were having trouble SETTING it. 

I took it out because there is no concept specifically of a single parent id. I would be ok with adding it back, but the agreed-upon definition is up to the user (aka is it the first ancestor? the last? ). So I wouldn't know how to define it to meet everyone's taste. 

Not opposed to having one, I just can't see how it would be correct rather than  parent_ids.last or parent_ids.first, etc.

**Comment by tjchambers** [comment]
The only way I can see it being a gem issue is if I missed a spot in removing USE of a getter for parent ID internally in the gem. Which is certainly possible.

**Comment by marcamillion** [comment]
Well in my case, what I am doing is trying to SET the value - but via the hidden field on the form partial. So what Rails seems to be doing is taking the value from the params and attempting to GET the value to stick it in the form so that it can be submitted when I create the new node....if that makes any sense, and when it goes to GET the value of the parent_id, that's where I imagine it is encountering that error.

In other words, for me to be able to set the `parent_id` attribute of the new `Node` object I have to set the value of `parent_id` in the form now so the ID can be sent with the request, and Rails couldn't find the value because there is no method `parent_id` on the model.

**Comment by tjchambers** [comment]
What I am struggling with is the use of the methods to access in the context of a multi-parent environment. Parent_id could be a list of all the parents, not just one, if the new one is going to be multi-parented. So you are really choosing - AFAICT - what "parent" you want to associate singly with the new node. Hence you should be populating it from a member of the list of parent_ids. 

Any implementation of a _singular_ parent_id requires a choice of how to define that it would seem in the context of a business decision. I purposely chose to drop methods rather than hardcode my interpretation (aka parent_id is ancestor_ids.last vs ancestor_ids.first vs ancestor_ids.min...)

That is distinct from how you fetch it into your hidden field or return it.

**Comment by marcamillion** [comment]
That's a good point. The truth is, I hadn't gotten that far yet. Before implementing the gem in my production app, I have re-built a mini-app that mimics a similar setup to my production app.

So I am just trying to figure out how your version works and seeing how your design decisions will impact how I implement this entire tree given my business logic context.

The first step was getting the basics up and running, and that's when I ran into this error....and I couldn't help but getting to the bottom of it. But now that you have mentioned it, that does make sense so I have to think through how I will do this.

Any suggestions on how to implement assigning multiple parents to a node?

**Comment by tjchambers** [comment]
Sorry - I missed responding to this. 

Not considering your issues of how to pass through a multi parent setting, I would say that parents_ids = "x" where x is an array of next-level parents you'd like to have would seem to push the right amount of work back on ancestry. Ancestry would then traverse those node ids and explode the node lineage, accumulate the results and populate the ancestry column as appropriate. 

That's my thinking.

**Comment by marcamillion** [comment]
I still haven't given up on needing single parent setter methods yet. Once I get the multiple parent setter functionality working, then maybe we can revisit it. It may just be a matter of updating the README to make it clear how to use this properly and how to get away from the single parent tree data structure.

See my other issue about multiple parent setters.

**Comment by tjchambers** [comment]
I am confused.

Isn't a single parent setter a simple case of multiparent setter with a single id?

**Comment by marcamillion** [comment]
You may be right...except that I can't get multiparent setter working even with a single id!

**Comment by tjchambers** [comment]
So parent=  does not work for you?

**Comment by tjchambers** [comment]
There is no parents= that is what we were discussing in the other thread.

**Comment by marcamillion** [comment]
Yep `parent=` does work.

Look forward to seeing a reply on the other thread!

**Comment by tjchambers** [comment]
Your comment about cleaning up the README is a good point. I didn't make an attempt to strip it of unused functionality or add/adjust the implemented functionality. I would entertain a pull request for that, and it may be valuable to you.

**Comment by marcamillion** [comment]
Yeh....once I grok everything more fully, I will do that.

I want to get it working properly first :)
